RFC-0022-Model-Partitioning-in-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,255 @@
+
+### Approach that enables more possibility of balanced partitioning
+
+* fx trace the graph, collect the statistics of each node
+* partition the graph using a bottom up approach: starting from treating each node as a partition, and merge them if it lowers the cost, use BFS to scan the merge order
+* The result will be an `fx.GraphModule` for each partition, the original module architecture is not preserved
+    * qual names of param might change
+
+Partition:
+Dict[fx.graphmodule, remotedevice]
+
+in order to apply the algorithm, we need to assign each top-level submodule with a proper cost, we calculate the cost based on memory and execution costs.
+
+
+### Cost calculation
+
+The cost of a submodule is defined as the following:
+
+```
+cost(m) = time_weight * execution_time(m) 
+            + memory_weight * memory_cost(m)
+
+memory_cost(m) = parameter cost * optimizer_multiplier 
+                    + activation_cost
+```
+
+where `memory_weight + time_weight = 1`. How to weigh between time and memory cost? Undecided, a heuristic number `time_weight=0.6`
+
+* @mrshenli: regarding balancing time between memory, I kind of feel we should prioritize time over memory. Because people usually use Pipeline parallel to accelerate training, and the training speed the ultimate goal. Balanced execution time has a directly impact on total pipeline makespan (suppose one phase is slower than others, then all other devices will need to wait for that phase). If the above assumption is correct, it looks like we should first try to balance time, and only try to balance memory when a shard of the optimally time-balanced model cannot fit in some devices. 
+
+Time > memory (memory constraint only, need to count in some buffer)
+
+How do we decide the `optimizer_multiplier` since different optimizers have different states maintained?
+
+We can use the param_scale mechanism which specify the scale for each parameters like in fairscale or torchgpipe. One disadvantage is that this `param_scale` is required from the user as a parameter, and user need to know the scale base on our documentation? Still seems a bad UX experience, but if we provide a default one, it’s likely not accurate.
+
+
+```
+    =========  =============  =========================================
+    Optimizer  `param_scale`  Internal State
+    =========  =============  =========================================
+    SGD        2--3           (momentum_buffer)
+    Adam       4--5           exp_avg, exp_avg_sq, (max_exp_avg_sq)
+    Adadelta   4              square_avg, acc_delta
+    Adagrad    3              sum
+    RMSprop    3--5           square_avg, (momentum_buffer), (grad_avg)
+    =========  =============  =========================================
+```
+
+
+Since we assigned the cost to each top-level module, we can do the partition, basic partition function will be like the below:
```

Comment:
Instead of hardcoding these constants, can we not run a few training iterations with the relevant optimizer?